### PR TITLE
Added sanctionsearch

### PR DESCRIPTION
**Pull request by k4u5h1k**
with individual search and id search. As mentioned in issue #149

### CODE REVIEW

File: core/util/sanctionsearch.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import json
+from bs4 import BeautifulSoup as bs
+
+class main:
+
+		def __init__(self, name, id=None, typ='individual', limit=15):
+			""" Sanctionsearch.ofac.treas.gov search engine
+					name     : name to search
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py:22]: id, typ, and limit need the comment

File: core/util/sanctionsearch.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+			self._name = name
+			self._type = typ
+			self._listid = 'ALL'
+			self._max = limit
+			self._id = id
+			self._json = ''
+			self.sanctionsearch = 'https://sanctionssearch.ofac.treas.gov/'
+			self._rows = []
+
+			if self._name is not None:
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: if self._name: isn't better?

Comment (saeeddhqan) [reply core/util/sanctionsearch.py]: self._data = [] if self._name else {}: maybe

File: core/util/sanctionsearch.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+			self.framework.verbose('Searching sanctionsearch...')
+			headers = {
+				'Content-Type': 'application/x-www-form-urlencoded',
+				}
+
+			try:
+				# First request to get viewstate from input tag
+				req = self.framework.request(url=self.sanctionsearch)
+				soup = bs(req.text,'html.parser')
+				viewstate = soup.find('input',
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: after except block. just request can be in try

File: core/util/sanctionsearch.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+					{'type':'hidden',
+					'name':'__VIEWSTATE',
+					'id':'__VIEWSTATE'})['value']
+
+				data = {'__VIEWSTATE':viewstate,
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+			headers = {
+				'Content-Type': 'application/x-www-form-urlencoded',
+				}
+
+			try:
+				# First request to get viewstate from input tag
+				req = self.framework.request(url=self.sanctionsearch)
+				soup = bs(req.text,'html.parser')
+				viewstate = soup.find('input',
+					{'type':'hidden',
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+						url=self.sanctionsearch,
+						method='POST',
+						data=data,
+						headers=headers)
+
+			except:
+				self.framework.error('[SANCTIONSEARCH] ConnectionError')
+				self.framework.error('Sanctionsearch is missed!')
+
+			soup = bs(req.text,'html.parser')
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+			table = soup.find('table', {'id':'gvSearchResults'})
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+
+		def id_crawl(self):
+			self._data = {}
+			url = self.sanctionsearch + f'Details.aspx?id={self._id}'
+			req = self.framework.request(url=url)
+			soup = bs(req.text,'html.parser')
+
+			# DETAILS
+			details_table = soup.find('table',
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+			if details_table is not None:
+				details_fields = map(lambda x: x.text, 
+					details_table.find_all('tr'))
+				details = list(filter(lambda x: len(x)>0, 
+					(''.join(details_fields)).split('\n')))
+
+				self._data['details'] = {}
+				for x,y in zip(details,details[1:]):
+					if x.endswith(':') and not y.endswith(':'):
+						self._data['details'][x]=y
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+			# IDENTIFICATION
+			id_table = soup.find('table',
+				{'id':'ctl00_MainContent_gvIdentification'})
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: modules/search/sanctionsearch.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	if name is None and id is None:
+		self.error('Either name or id is required')
+		return
+	elif name is not None and id is not None:
+		self.error('Only specify either name or id not both')
+		return
+	elif name is not None:
+		NAMESEARCH=True
+	
+	run = self.sanctionsearch(name=name,id=id,limit=limit)
```
Comment (saeeddhqan) [inline modules/search/sanctionsearch.py]: space error

File: modules/search/sanctionsearch.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+	if NAMESEARCH:
+		run.name_crawl()
+		output = {'results': []}
+		links = run.data
+
+		for item in links:
+			output['results'].append(item)
+
+		self.save_gather(output, 'search/sanctionsearch', name, output=self.options['output'])
```
Comment (saeeddhqan) [inline modules/search/sanctionsearch.py]: don't call it twice. use a toggle variable for name and id. `base = name if .. else id`

**Review by saeeddhqan — COMMENTED** [review]
Great. Solve the notes

File: core/util/sanctionsearch.py (hunk 2 above)
Comment (k4u5h1k) [reply core/util/sanctionsearch.py:36]: > if self._name: isn't better?

No thats inaccurate.
Everything else I shall fix

File: core/util/sanctionsearch.py (hunk 3 above)
Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: There are two requests.

Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: So two try except blocks?

**Comment by k4u5h1k** [comment]
All done I guess

**Comment by k4u5h1k** [comment]
All done I guess

File: core/util/sanctionsearch.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+						headers=headers)
+			except:
+				self.framework.error('[SANCTIONSEARCH] ConnectionError')
+				self.framework.error('Sanctionsearch is missed!')
+				return
+
+			soup = bs(req.text, 'html.parser')
+			table = soup.find('table', {'id': 'gvSearchResults'})
+
+			if table is not None:
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: Again, use one-line if-else: self._rows = .. if .. else ..

File: core/util/sanctionsearch.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+				self._rows = table.find_all('tr')
+			else:
+				self._rows = None
+
+			if self._rows is not None:
+				for count, row in enumerate(self._rows):
+					if count>=self._max:
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+					self._data.append({
+						'name': name,
+						'address': address,
+						'link': self.sanctionsearch+link
+						})
+
+
+		def id_crawl(self):
+			self._data = {}
+			url = self.sanctionsearch + f'Details.aspx?id={self._id}'
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: f".."

File: core/util/sanctionsearch.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+			if id_table is not None:
+				id_th = list(map(lambda x: x.text,
+					id_table.find_all('th', {'class': 'borderline'})))
+				id_td = list(map(lambda x: x.text,
+					id_table.find_all('td')))
+
+				self._data['identification'] = {}
+				for i in range(len(id_th)):
+					if len(id_td[i].strip())>0:
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+			if al_table is not None:
+				al_th = list(map(lambda x: x.text,
+					al_table.find_all('th', {'class': 'borderline'})))
+				al_td = list(map(lambda x: x.text,
+					al_table.find_all('td')))
+
+				self._data['Aliases'] = {}
+				for i in range(len(al_th)):
+					if len(al_td[i].strip())>0:
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: core/util/sanctionsearch.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+				address_rows =list(filter(lambda x:len(''.join(x.text).strip())>0, 
+					address_table.find_all('tr')))[1:]
+
+				self._data['addresses'] = {}
+				for i,row in enumerate(address_rows, 1):
+					cols = list(map(lambda x: x.text.strip(),
+						row.find_all('td')))
+					self._data['addresses'][str(i)] = {}
+					for j in range(len(cols)):
+						if len(cols[j])>0:
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: space error

File: modules/search/sanctionsearch.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+
+meta = {
+        'name': 'SanctionSearch',
+        'author': 'Kaushik',
+        'version': '0.1',
+        'description': 'SanctionSearch is an international list of blacklisted, dangerous\
+		personnel and entities.',
+        'sources': ('sanctionsearch'),
+        'options': (
+		('name', None, False, 'name to search', '-n', 'store', str),
```
Comment (saeeddhqan) [inline modules/search/sanctionsearch.py]: query is better

File: modules/search/sanctionsearch.py [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+	return output
+
+def module_run(self):
+	output = module_api(self)
+	if NAMESEARCH:
+		output = output['results']
+		for name in output:
+			print()
+			self.output(name['name'])
+			if len(name['address'].strip())>0:
```
Comment (saeeddhqan) [inline modules/search/sanctionsearch.py]: space error

**Comment by k4u5h1k** [comment]
pep8 says not to put spaces around operators in if conditions no?

File: core/util/sanctionsearch.py (hunk 8 above)
Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: parameter '=' should never be surrounded by spaces

**Comment by k4u5h1k** [comment]
done

File: core/util/sanctionsearch.py [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+					self._data.append({
+						'name': name,
+						'address': address,
+						'link': self.sanctionsearch+link
+						})
+
+
+		def id_crawl(self):
+			self.framework.verbose('Searching sanctionsearch...')
+			url = self.sanctionsearch + f"Details.aspx?id={self._id}"
```
Comment (saeeddhqan) [inline core/util/sanctionsearch.py]: add self.sanctionsearch to the quotes.

File: core/util/sanctionsearch.py (hunk 8 above)
Comment (saeeddhqan) [reply core/util/sanctionsearch.py]: If you have a good source for it, we can change it comprehensively.

File: core/util/sanctionsearch.py (hunk 21 above)
Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: oh yes

File: core/util/sanctionsearch.py (hunk 8 above)
Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: sure give me a sec

Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: - https://www.flake8rules.com/rules/E251.html
- https://stackoverflow.com/questions/8853063/pep-8-why-no-spaces-around-in-keyword-argument-or-a-default-parameter-value

Comment (saeeddhqan) [reply core/util/sanctionsearch.py]: This is for function parameters, not the variables.

Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: > parameter '=' should never be surrounded by spaces

yes i was talking about function parameters

Comment (k4u5h1k) [reply core/util/sanctionsearch.py]: my bad I thought you were talking about that

**Comment by saeeddhqan** [comment]
Okay, looks good.

**Comment by z3y50n** [comment]
When I run `sanctionsearch` I get this.

image

Also when I search for id=40 I get results in the website
image
But on the tool I don't get anything
image

File: core/util/sanctionsearch.py [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import json
+from bs4 import BeautifulSoup as bs
+
+class main:
+
```
Comment (z3y50n) [inline core/util/sanctionsearch.py:19]: Remove this empty line

File: core/util/sanctionsearch.py [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import json
```
Comment (z3y50n) [inline core/util/sanctionsearch.py]: Not used.

File: modules/search/sanctionsearch.py [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+	
+	run = self.sanctionsearch(query=query, id=id, limit=limit)
+
+	if NAMESEARCH:
+		output_param = query
+		run.name_crawl()
+		output = {'results': []}
+		links = run.data
+		for item in links:
+			output['results'].append(item)
```
Comment (z3y50n) [inline modules/search/sanctionsearch.py]: You can use `extend`

File: core/util/sanctionsearch.py [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+
+				self._data['addresses'] = {}
+				for i,row in enumerate(address_rows, 1):
+					cols = list(map(lambda x: x.text.strip(),
+						row.find_all('td')))
+					self._data['addresses'][str(i)] = {}
+					for j in range(len(cols)):
+						if len(cols[j]) > 0:
+							self._data['addresses'][str(i)][address_th[j]] = cols[j]
+
```
Comment (z3y50n) [inline core/util/sanctionsearch.py:157]: Separate functions with one empty line only

File: core/util/sanctionsearch.py [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+					address = row.find_all('td')[1].text
+					link = row.find('a')['href']
+
+					self._data.append({
+						'name': name,
+						'address': address,
+						'link': self.sanctionsearch+link
+						})
+
+
```
Comment (z3y50n) [inline core/util/sanctionsearch.py]: Separate functions with one empty line

**Comment by k4u5h1k** [comment]
> When I run `sanctionsearch` I get this.
> 
> image
> 
> Also when I search for id=40 I get results in the website
> image
> But on the tool I don't get anything
> image

Issue 1:
Oh wow I just noticed this was happening in my other modules as well.
The problem was caused because of not having a comma at the end of the sources tuple.
So, any module with 
`'sources': ('name')` has the problem
`'sources': ('name',)` solves it 

(why is this happening?)
EDIT (5 months later) : A string in python is a tuple with that string as the first element. Thats why this was happening.  
Screenshot 2021-08-18 at 10 22 29 AM


Issue 2:
That is not accurate their ids are not 40. 
To confirm that you may click on the first result and observe the id in the url which is
`https://sanctionssearch.ofac.treas.gov/Details.aspx?id=29522`

Thanks for the review!

File: modules/search/sanctionsearch.py (hunk 24 above)
Comment (k4u5h1k) [reply modules/search/sanctionsearch.py]: Each item is a dict of a row from the site, so they must be compartmentalized. Perhaps i modified the code off of an existing module and the links variable name was left unchanged, I'll change that

**Comment by k4u5h1k** [comment]
Is there anything left?

**Comment by saeeddhqan** [comment]
I'll check it out and merge it. You're done!
Thank you.

**Pull request merged**